src/ripple/core/JobCounter.h

```diff
@@ -0,0 +1,188 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2017 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_JOB_COUNTER_H_INCLUDED
+#define RIPPLE_CORE_JOB_COUNTER_H_INCLUDED
+
+#include <ripple/core/Job.h>
+#include <boost/optional.hpp>
+#include <condition_variable>
+#include <mutex>
+
+namespace ripple {
+
+// A class that does reference counting for Jobs.  The reference counting
+// allows a Stoppable to assure that all child Jobs in the JobQueue are
+// completed before the Stoppable declares itself stopped().
+class JobCounter
+{
+private:
+    std::mutex mutable mutex_ {};
+    std::condition_variable allJobsDoneCond_ {};  // guard with mutex_
+    bool waitForJobs_ {false};                    // guard with mutex_
+    int jobCount_ {0};                            // guard with mutex_
+
+    // Manage the count.  If we're stopping and the count drops to zero
+    // notify allJobsDoneCond_.
+    void adjustCount (int upDn)
+    {
+        std::lock_guard<std::mutex> lock {mutex_};
+
+        // Count the job.  Notify if we're stopping and all jobs are done.
+        jobCount_ += upDn > 0 ? 1 : -1;
+        if (waitForJobs_ && jobCount_ == 0)
+            allJobsDoneCond_.notify_all();
+    }
```

Comment:
Since `adjustCount` is private and `jobCount_` can only be zero when called from a CountedJob dtor, we unnecessarily check to notify with ctors. A simpler solution to consider is to define prefix increment/decrement overloads instead of `adjustCount`.

`    JobCounter& JobCounter::operator++()
    {
        std::lock_guard<std::mutex> lock {mutex_};

        ++jobCount_;
        return *this;
    }

    JobCounter& JobCounter::operator--()
    {
        std::lock_guard<std::mutex> lock {mutex_};

        // Count the job.  Notify if we're stopping and all jobs are done.
        if (--jobCount_ == 0 && waitForJobs_)
            allJobsDoneCond_.notify_all();
        return *this;
    }`